### PR TITLE
[Pruning algo] Filter counting fix for linear layers

### PR DESCRIPTION
**Pull request by daniil-lyakhov**
### Changes
THIS WILL CHANGE PRUNING LEVEL OF MODELS WHICH CONTAINS PRUNABLE LINEAR LAYERS, PRUNED BY FLOPS

1) Weights and flops calculation for pruned graphs with closing linear layer was fixed by `sparse_multiplier` param introduction
2) Groups with different amount of channels / output elements are filtering during `pruing_node_selector` call
3) Batched linear nodes are filtering during  `pruing_node_selector` call
4) NAS: Wrong weights / flops calculation for models with depthwise convs was fixed
5) NAS: `purning_node_selector` and shape pruning metatypes was aligned
6) Weights / flops calculation was refactored

### Reason for changes

1) Wrong flops/weights calculation in case of closing linear layer. `sparse_multiplier` is propagating through net and inside an input mask of a closing operation it's equal to amount of input channels of conv layer / input elements of linear layer corresponded to one output channel / output element of producer. 
For example in a case bellow conv_1 mask producer and conv_2 mask producer have `sparse_multiplier` equal to `H*W`
image

2)

3) Bathced linear is an linear which output edge with >= 2 dimentions. Example, in which linear will affect feature map of closing conv instead of channels in case of pruning, which is not right.
image

### Related tickets

86889

### CODE REVIEW

File: nncf/common/pruning/utils.py [hunk 1]
```diff
@@ -278,8 +274,6 @@ def count_flops_and_weights_per_node(graph: NNCFGraph,
     Counts the number weights and FLOPs per node in the model for convolution and fully connected layers.
 
     :param graph: NNCFGraph.
-    :param input_shapes: Dictionary of input dimension shapes for convolutions and
-        fully connected layers. E.g {node_name: (height, width)}
     :param output_shapes: Dictionary of output dimension shapes for convolutions and
         fully connected layers. E.g {node_name: (height, width)}
```
Comment (sooahleex) [inline nncf/common/pruning/utils.py:278]: It seems that `output_shapes` variable also no longer needed.

File: nncf/common/pruning/utils.py [hunk 2]
```diff
@@ -238,8 +237,6 @@ def count_flops_and_weights(graph: NNCFGraph,
     Counts the number weights and FLOPs in the model for convolution and fully connected layers.
 
     :param graph: NNCFGraph.
-    :param input_shapes: Dictionary of input dimension shapes for convolutions and
-        fully connected layers. E.g {node_name: (height, width)}
     :param output_shapes: Dictionary of output dimension shapes for convolutions and
         fully connected layers. E.g {node_name: (height, width)}
```
Comment (sooahleex) [inline nncf/common/pruning/utils.py:241]: If `count_flops_and_weights_per_node` not need output_shapes, this function also did not use `output_shapes`.

**Review by sooahleex — COMMENTED** [review]
The `output_shape(or out_shape)` also seems unnecessary and can be replaced with `out_features`. Is there a reason why you left `output_shape`?

File: nncf/common/pruning/utils.py (hunk 1 above)
Comment (sooahleex) [reply nncf/common/pruning/utils.py:278]: `output_shapes` is needed for computing flops
https://github.com/openvinotoolkit/nncf/blob/d5e9edf8f335fc6d97825ddc07cfbd03594defd2/nncf/common/pruning/utils.py#L321

**Comment by sooahleex** [comment]
Looks good to me

File: tests/torch/pruning/filter_pruning/test_flops_pruning.py [hunk 3]
```diff
@@ -84,12 +84,12 @@ def test_init_params_for_flops_calculation(model, ref_params):
     ("model", "all_weights", "pruning_flops_target", "ref_full_flops", "ref_current_flops", "ref_sizes"),
     (
         (PruningTestWideModelConcat, False, 0.4, 671154176, 399057920, [328, 656, 656]),
-        (PruningTestWideModelEltwise, False, 0.4, 268500992, 160773120, [360, 720, 720]),
-        (PruningTestWideModelConcat, True, 0.4, 671154176, 402513920, [380, 647, 648]),
-        (PruningTestWideModelEltwise, True, 0.4, 268500992, 161043328, [321, 755, 755]),
-        (PruningTestModelSharedConvs, True, 0.4, 461438976, 276594768, [361, 809]),
-        (PruningTestModelSharedConvs, False, 0.4, 461438976, 275300352, [384, 768]),
-        (GroupedConvolutionModel, False, 0.0, 11243520, 11243520, [])
+        #(PruningTestWideModelEltwise, False, 0.4, 268500992, 160773120, [360, 720, 720]),
```
Comment (alexsu52) [inline tests/torch/pruning/filter_pruning/test_flops_pruning.py:87]: Could you briefly explain why you commented out these tests?

Comment (daniil-lyakhov) [reply tests/torch/pruning/filter_pruning/test_flops_pruning.py:87]: This pr is still WIP, I need to update all this values, sorry for inconvenience

File: nncf/common/pruning/symbolic_mask.py [hunk 4]
```diff
@@ -27,17 +58,22 @@ class SymbolicMask(NNCFTensor):
     symbolic mask propagation by SymbolicMaskProcessor.
     """
 
-    def __init__(self, dimension: int, mask_producers: List[int] = None):
+    def __init__(self, dimension: int, mask_producers: Union[int, List[SymbolicMaskProducer]] = None):
```
Comment (ljaljushkin) [inline nncf/common/pruning/symbolic_mask.py:61]: In which cases are mask producers equal to int?

File: nncf/tensorflow/pruning/filter_pruning/algorithm.py [hunk 5]
```diff
@@ -520,25 +470,23 @@ def _calculate_filters_importance_in_group(self, group: Cluster[PrunedLayerInfo]
...
-                full_input_channels=self._layers_in_channels,
-                full_output_channels=self._layers_out_channels,
-                pruning_groups_next_nodes=self._next_nodes)
-
-        self.current_filters_num = count_filters_num(self._original_graph,
-                                                     op_metatypes=GENERAL_CONV_LAYER_METATYPES +
-                                                     LINEAR_LAYER_METATYPES,
-                                                     output_channels=tmp_out_channels)
+        tmp_in_channels, tmp_out_channels = \
+            self._shape_pruning_proc.calculate_in_out_channels_by_masks(
```
Comment (ljaljushkin) [inline nncf/tensorflow/pruning/filter_pruning/algorithm.py:474]: `calculate_in_out_channels_by_masks` calculate something that is needed for calculation of flops via `weights_flops_calc` only. Why not hide this calculation in the `weight_flops_calc` then, or maybe in another entity?

Unfortunately, `ShapePruninigProcessor` and `WeightsFlopsCalculator` don't encapsulate a lot - just metatypes and pruning types. 
It would be super-great to have a class with a simple and minimal interface, like `calc_flops` and `set_graph`.

Comment (daniil-lyakhov) [reply nncf/tensorflow/pruning/filter_pruning/algorithm.py:474]: 1) I made separate entities because elastic width needs only `ShapePruninigProcessor` and multi elasticity handler needs only `WeightsFlopsCalculator`. Maybe it worth to have one new entity when you need both classes which compose `ShapePruninigProcessor`  and `WeightsFlopsCalculator`, let's discuss it.

2) Before I dive into NAS algo the classes kept all needed information inside. But when I found that NAS can change not just graph, but pruning groups, I decided to keep only things which will not be changed during any algorithm. 

There are two options: keep params and than update it or feed params and explicitly ask algorithm to manage such parameters. Param feeding is more explicit way, so I vote for it

Comment (daniil-lyakhov) [reply nncf/tensorflow/pruning/filter_pruning/algorithm.py:474]: Other thing I was thinking is that `next_nodes` and `pruning_groups` always should be in one place, this is thoughts for future common pruning algo

File: nncf/common/pruning/symbolic_mask.py (hunk 4 above)
Comment (daniil-lyakhov) [reply nncf/common/pruning/symbolic_mask.py:61]: Im `symbolic_mask_propagation`, just sugar to not initialize symbolic mask like
```
mask = SymbolicMask(dimension=dim, mask_producers=MaskProducer(producer))
```

it works like that instead:
```
mask = SymbolicMask(dimension=dim, mask_producers=producer)
```

**Review by ljaljushkin — APPROVED** [review]

File: tests/common/pruning/test_pruning_operations.py [hunk 6]
```diff
@@ -3,6 +3,11 @@
 
 from functools import partial
 
+from tests.common.pruning import dummy_types
+from nncf.common.pruning.tensor_processor import NNCFPruningBaseTensorProcessor
+from tests.common.pruning.tensor import NPNNCFTensor
+from tests.common.pruning.tensor import NPNNCFTensorProcessor
+from nncf.common.graph.layer_attributes import ReshapeLayerAttributes
```
Comment (alexsu52) [inline tests/common/pruning/test_pruning_operations.py:10]: Please, re-sort imports.

File: tests/common/pruning/test_symbolic_mask_processor.py [hunk 7]
```diff
@@ -0,0 +1,132 @@
...
+    masks = [SymbolicMask(5 if all_close else i, producer) for i, producer in enumerate(masks_producers)]
+    result = SymbolicMaskProcessor.elementwise_mask_propagation(masks)
+    assert {p.id for p in result.mask_producers} == set(range(3))
+    if not all_close:
+        assert isinstance(result, AmbiguousSymbolicMask)
+    else:
+        assert result.shape[0] == 5
+
+
+
```
Comment (alexsu52) [inline tests/common/pruning/test_symbolic_mask_processor.py:99]: Please, remove extra blank line

File: tests/tensorflow/pruning/helpers.py [hunk 8]
```diff
@@ -174,3 +174,57 @@ def get_split_test_model(input_shape):
     x = tf.concat([y1, x], -1, name='tf_concat')
     outputs = conv4(x)
     return tf.keras.Model(inputs=inputs, outputs=outputs)
+
```
Comment (alexsu52) [inline tests/tensorflow/pruning/helpers.py:177]: Please, add a blank line

**Review by alexsu52 — COMMENTED** [review]
Thanks for your contribution! Looks good overall.

File: nncf/common/pruning/weights_flops_calculator.py [hunk 9]
```diff
@@ -0,0 +1,161 @@
...
+
+        :param conv_op_metatypes: List of metatypes defining convolution operations.
+        :param linear_op_metatypes: List of metatypes defining linear/fully connected operations.
+        """
+        self._conv_op_metatypes = conv_op_metatypes
+        self._linear_op_metatypes = linear_op_metatypes
+
+    def count_flops_and_weights(self,
+                                graph: NNCFGraph,
+                                output_shapes: Dict[NNCFNodeName, int],
```
[the remark this answers is not in the document]
Comment (daniil-lyakhov) [reply nncf/common/pruning/weights_flops_calculator.py:48]: I've checked, this attribute is required for flops and weights calculation. Mainly to multiply amount of flops on output elements of conv and to multiply amount of flops on amount of matrices in linear layer

File: nncf/common/pruning/utils.py [hunk 10]
```diff
@@ -42,6 +36,18 @@ def is_grouped_conv(node: NNCFNode) -> bool:
            and node.layer_attributes.groups != 1
 
 
+def is_batched_linear(node: NNCFNode, graph: NNCFGraph) -> bool:
```
[the remark this answers is not in the document]
Comment (daniil-lyakhov) [reply nncf/common/pruning/utils.py:45]: Done, please check

File: nncf/common/pruning/shape_pruning_processor.py [hunk 11]
```diff
@@ -0,0 +1,214 @@
...
+            num_of_sparse_elems = sparse_elements_counter(layer_name)
+
+            self.prune_cluster_shapes(cluster=group, pruned_elems=num_of_sparse_elems,
+                                      input_channels=tmp_in_channels,
+                                      output_channels=tmp_out_channels,
+                                      pruning_groups_next_nodes=pruning_groups_next_nodes)
+
+        return tmp_in_channels, tmp_out_channels
+
+    class NextNodeInfo:
```
[the remark this answers is not in the document]
Comment (daniil-lyakhov) [reply nncf/common/pruning/shape_pruning_processor.py]: Fixed this, please check

File: nncf/common/pruning/operations.py [hunk 12]
```diff
@@ -320,18 +326,13 @@ def _is_flatten(node: NNCFNode) -> bool:
...
-    def _is_not_mixing_dim(node: NNCFNode, graph) -> bool:
+    def _is_not_mixing_dim(node: NNCFNode) -> bool:
         input_shape = node.layer_attributes.input_shape
-
-        output_edges = graph.get_output_edges(node)
-        if not output_edges:
-            return True
         output_shape = node.layer_attributes.output_shape
 
+        # TODO: fix according to CVS-90976
```
[the remark this answers is not in the document]
Comment (daniil-lyakhov) [reply nncf/common/pruning/operations.py]: Done

File: nncf/common/pruning/node_selector.py [hunk 13]
```diff
@@ -194,7 +197,33 @@ def _get_multiforward_nodes(self, graph: NNCFGraph) -> List[List[NNCFNode]]:
             ret[node.layer_name].append(node)
         return [ret[module_identifier] for module_identifier in ret if len(ret[module_identifier]) > 1]
 
-    def _pruning_dimensions_analysis(self, graph, can_prune_after_check) -> Dict[int, PruningAnalysisDecision]:
+    def _pruning_dimensions_analysis(self, graph: NNCFGraph,
+                                     pruned_nodes_clusterization: Clusterization,
+                                     can_prune_after_check: Dict[int, PruningAnalysisDecision]) ->\
+                                     Dict[int, PruningAnalysisDecision]:
+        """
+        Checks all nodes that were marked as prunable after the model analysis and compatibility check vs.
```
[the remark this answers is not in the document]
Comment (daniil-lyakhov) [reply nncf/common/pruning/node_selector.py]: Done, please check

**Review by alexsu52 — APPROVED** [review]
